receiver/awsfirehosereceiver/unmarshaler/cwmetricstream/metricsbuilder.go

```diff
@@ -0,0 +1,157 @@
+
+	"go.opentelemetry.io/collector/model/pdata"
+	conventions "go.opentelemetry.io/collector/model/semconv/v1.5.0"
+)
+
+const (
+	attributeAWSCloudWatchMetricStreamName = "aws.cloudwatch.metric_stream_name"
+	dimensionInstanceID                    = "InstanceId"
+	namespaceDelimiter                     = "/"
+)
+
+type resourceMetricsBuilder struct {
+	metricStreamName string
+	accountID        string
+	region           string
+	namespace        string
+	metricBuilders   map[string]*metricBuilder
+}
+
+func newResourceMetricsBuilder(metricStreamName, accountID, region, namespace string) *resourceMetricsBuilder {
+	return &resourceMetricsBuilder{
+		metricStreamName: metricStreamName,
+		accountID:        accountID,
+		region:           region,
+		namespace:        namespace,
+		metricBuilders:   make(map[string]*metricBuilder),
+	}
+}
+
+// AddMetric adds a metric to one of the metric builders based on
+// the key generated for each.
+func (rmb *resourceMetricsBuilder) AddMetric(metric cWMetric) {
+	metricKey := rmb.toMetricKey(metric)
+	mb, ok := rmb.metricBuilders[metricKey]
+	if !ok {
+		mb = newMetricBuilder(metric.MetricName, metric.Unit)
+		rmb.metricBuilders[metricKey] = mb
+	}
+	mb.AddDataPoint(metric)
+}
+
+// Build creates the pdata.ResourceMetrics from the added metrics.
+func (rmb *resourceMetricsBuilder) Build() pdata.ResourceMetrics {
+	rm := pdata.NewResourceMetrics()
+	ilm := rm.InstrumentationLibraryMetrics().AppendEmpty()
+	rmb.toResource().CopyTo(rm.Resource())
+	for _, mb := range rmb.metricBuilders {
+		mb.Build().CopyTo(ilm.Metrics().AppendEmpty())
+	}
+	return rm
+}
+
+// toResource creates a pdata.Resource from the fields in the resourceMetricsBuilder.
+// Splits the namespace into service.namespace/service.name if prepended by AWS/.
+func (rmb *resourceMetricsBuilder) toResource() pdata.Resource {
+	resource := pdata.NewResource()
+	attributes := resource.Attributes()
+	attributes.InsertString(conventions.AttributeCloudProvider, conventions.AttributeCloudProviderAWS)
+	attributes.InsertString(conventions.AttributeCloudAccountID, rmb.accountID)
+	attributes.InsertString(conventions.AttributeCloudRegion, rmb.region)
+	splitNamespace := strings.SplitN(rmb.namespace, namespaceDelimiter, 2)
+	if len(splitNamespace) == 2 && strings.EqualFold(splitNamespace[0], conventions.AttributeCloudProviderAWS) {
+		attributes.InsertString(conventions.AttributeServiceNamespace, splitNamespace[0])
+		attributes.InsertString(conventions.AttributeServiceName, splitNamespace[1])
+	} else {
+		attributes.InsertString(conventions.AttributeServiceName, rmb.namespace)
+	}
+	attributes.InsertString(attributeAWSCloudWatchMetricStreamName, rmb.metricStreamName)
+	return resource
+}
+
+// toMetricKey creates a key based on the metric name and dimensions to
+// keep metrics with different dimensions separate.
+func (rmb *resourceMetricsBuilder) toMetricKey(metric cWMetric) string {
+	return fmt.Sprintf("%s::%v", metric.MetricName, metric.Dimensions)
```

Comment:
```suggestion
	return metric.MetricName
```

I think this should simply be keyed on metric name.  Different sets of dimensions for the metric can be represented as data points within the metric.

Reply:
If they're in the same metric, wouldn't they be aggregated together? Based on the CloudWatch [documentation](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/cloudwatch_concepts.html#Dimension), 

> Because dimensions are part of the unique identifier for a metric, whenever you add a unique name/value pair to one of your metrics, you are creating a new variation of that metric.

> CloudWatch treats each unique combination of dimensions as a separate metric, even if the metrics have the same metric name.

I believe we wouldn't want to aggregate metrics with the same name, but different dimensions since that could cause incorrect sums.

Reply:
I think CloudWatch metrics are analogous to OTel's [timeseries model](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/datamodel.md#timeseries-model), whereas the pdata metric is of the [OTLP/stream model](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/datamodel.md#opentelemetry-protocol-data-model).

Here, we should probably map the CW metrics to [Summary data points](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/datamodel.md#summary-legacy).  The metric level would contain just the [name, description, and unit](https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/metrics/v1/metrics.proto#L87), while the data point would [contain attributes, timestamps, and values](https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/metrics/v1/metrics.proto#L89-L90).